target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,506 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * Realtek RTL83XX clock driver
+ * Copyright (C) 2022 Markus Stockhausen
+ *
+ * This driver provides basic clock support for RTL838X and RTL8389X SoC.
+ * Currently CPU, memory and LXB clock information can be accessed. 
+ * To make use of the driver add the following clocks section to the DT.
+ *
+ * #include <dt-bindings/clock/rtl83xx-clk.h>
+ * clocks {
+ *   osc: oscillator {
+ *     compatible = "fixed-clock";
+ *     #clock-cells = <0>;
+ *     clock-frequency = <25000000>;
+ *   };
+ *   ccu: clock-controller {
+ *     compatible = "realtek,rtl83xx-clock";
+ *     #clock-cells = <1>;
+ *     clocks = <&osc>;
+ *     clock-names = "ref_clk";
+ *   };
+ * };
+ *
+ * Afterwards one can access/identify the clocks in the DT with
+ * <&ccu CLK_CPU>, <&ccu CLK_MEM> or <&ccu CLK_LXB>. Additionally
+ * the clocks can be used inside the kernel with
+ *
+ * cpu_clk = clk_get(NULL, "cpu_clk");
+ * mem_clk = clk_get(NULL, "mem_clk");
+ * lxb_clk = clk_get(NULL, "lxb_clk");
+ *
+ */
+
+#include <asm/cacheflush.h>
+#include <linux/clk.h>
+#include <linux/clk-provider.h>
+#include <linux/clkdev.h>
+#include <linux/cpu.h>
+#include <linux/delay.h>
+#include <linux/io.h>
+#include <linux/ioport.h>
+#include <linux/of_address.h>
+#include <linux/slab.h>
+
+#include <dt-bindings/clock/rtl83xx-clk.h>
+#include <asm/mach-rtl838x/mach-rtl83xx.h>
+
+extern struct rtl83xx_soc_info soc_info;
+
+/*
+ * some hardware specific definitions
+ */
+
+#define SOC_RTL838X 	0
+#define SOC_RTL839X	1
+#define SOC_COUNT	2
+
+#define REG_CTRL0	0
+#define REG_CTRL1	1
+#define REG_COUNT	2
+
+#define CLK_COUNT	3
+
+#define OSC_RATE	25000000
+
+static const int rtl_regs[SOC_COUNT][REG_COUNT][CLK_COUNT] = {
+	{
+		{ RTL838X_PLL_CPU_CTRL0, RTL838X_PLL_MEM_CTRL0, RTL838X_PLL_LXB_CTRL0 },
+		{ RTL838X_PLL_CPU_CTRL1, RTL838X_PLL_MEM_CTRL1, RTL838X_PLL_LXB_CTRL1 },
+	}, {
+		{ RTL839X_PLL_CPU_CTRL0, RTL839X_PLL_MEM_CTRL0, RTL839X_PLL_LXB_CTRL0 },
+		{ RTL839X_PLL_CPU_CTRL1, RTL839X_PLL_MEM_CTRL1, RTL839X_PLL_LXB_CTRL1 },
+	}
+};
+
+#define REG_SET(_rate, _ctrl0, _ctrl1, _misc)	\
+	{					\
+		.rate = _rate,			\
+		.ctrl0 = _ctrl0,		\
+		.ctrl1 = _ctrl1,		\
+		.misc = _misc			\
+	}
+
+struct rtl_reg_set {
+	unsigned int rate;
+	unsigned int ctrl0;
+	unsigned int ctrl1;
+	unsigned int misc;
+};
+
+static const struct rtl_reg_set rtl838x_cpu_reg_set[] = {
+	REG_SET(300000000, 0x045c8, 0x1414530e, 0x2d20cfb8),
+	REG_SET(500000000, 0x04748, 0x0c14530e, 0x2da0cfb8)
```

Comment:
Are these the pre-combined values of all those little register fields you had in the previous version?

Reply:
yep. These are the hardware magic values. using constants is better than some ugly code combination.